src/main/java/org/prebid/server/activity/infrastructure/ActivityInfrastructure.java

```diff
@@ -33,14 +34,24 @@ private static void validate(Map<Activity, ActivityController> activitiesControl
         }
     }
 
-    public boolean isAllowed(Activity activity, ActivityInvocationPayload activityInvocationPayload) {
-        debug.emitActivityInvocation(activity, activityInvocationPayload);
-        final boolean result = activitiesControllers.get(activity).isAllowed(activityInvocationPayload);
-        debug.emitActivityInvocationResult(activity, activityInvocationPayload, result);
+    public boolean isAllowed(Activity activity, ActivityInvocationPayload payload) {
+        validatePayloadType(payload);
+
+        final CompositeActivityInvocationPayload compositePayload = (CompositeActivityInvocationPayload) payload;
+        debug.emitActivityInvocation(activity, compositePayload);
+        final boolean result = activitiesControllers.get(activity).isAllowed(compositePayload);
+        debug.emitActivityInvocationResult(activity, compositePayload, result);
 
         return result;
     }
 
+    private static void validatePayloadType(ActivityInvocationPayload activityInvocationPayload) {
+        if (!(activityInvocationPayload instanceof CompositeActivityInvocationPayload)) {
+            throw new IllegalArgumentException(
+                    "Invalid payload type. Please, consider to use 'ActivityPayload.builder()'.");
+        }
+    }
```

Comment:
We can get rid of this invariant check. Instead of `ActivityPayload.builder()` we can implement `ActivityPayload.forComponent(ComponentType componentType, String componentName)`